Supported `file` scheme without netloc.

KOLANICH opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

Fixes #2074

**Description of the changes being introduced by the pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

KOLANICH committed:
Supported schemes without netloc.

Signed-off-by: KOLANICH <KOLANICH@users.noreply.github.com>

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -118,7 +120,7 @@ def _get_session(self, url: str) -> requests.Session:
         parsed_url = parse.urlparse(url)
 
-        if not parsed_url.scheme or not parsed_url.hostname:
+        if not parsed_url.scheme or (not parsed_url.hostname and parsed_url.scheme not in self.SCHEMES_WITHOUT_HOSTNAME):
```
this is unlikely to pass our linter (line looks too long). you can use `tox` to run linters and tests locally.

That said, I wonder if the check is even useful? IIRC it was added just to replace some earlier URL sanity checks, and it's certainly not "real" URL validation: The intent is only to make sure the session index lookup a few lines later works as intended and it seems to me that it does... I think removing the host name check (or even the whole if-clause) may be fine.

jku reviewed (changes requested):
Yeah I agree with the intent, this seems good.

I believe this will not pass linting test though, and I think the whole check could maybe be removed (see comment).

If you do modify the commit, please consider adding a commit message describing why this is done. I assume the reason is something like _"This enables using ngclient with a repository accessible via file:// URLs"_ ?

jku commented:
oh signoff is also missing, please check "DCO" here: https://github.com/theupdateframework/python-tuf/blob/develop/docs/CONTRIBUTING.rst#dco

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -43,7 +43,7 @@ def __init__(self) -> None:
         # hosts-scheme combinations to minimize subtle security issues.
         # Some cookies may not be HTTP-safe.
-        self._sessions: Dict[str, requests.Session] = {}
+        self._sessions: Dict[Tuple[str, Optional[str]], requests.Session] = {}
```
nitpick: no need for Optional: hostname is guaranteed to be a string (it may be an empty string but it's not None).

jku commented:
DCO failure is likely because your commit author email and the signoff email are not the same. I assume using the noreply address works (if that's what you wanted to do) as long as it's used consistently but I have never tried myself

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -6,7 +6,7 @@
 
 import logging
-from typing import Dict, Iterator
+from typing import Dict, Iterator, Tuple, Optional
```
linter is complaining about the order here but `Optional` is also unused: could be removed entirely.

Sorry, we have a lot of lint checks -- the code used to be a mishmash of styles and we decided being strict is better than dealing with that...

jku commented:
> nitpick: no need for Optional: hostname is guaranteed to be a string (it may be an empty string but it's not None).

Argh, it looks like the typeshed annotations for requests disagree (they look buggy to me, but :shrug:). Very sorry, that is entirely on me.

I'm fine with your original use of "Optional" or alternatively this probably works too as a oneline change:
`session_index = (parsed_url.scheme, parsed_url.hostname or "")`

jku commented:
If you want to run the linter locally, `tox -e lint` works once you've [installed](https://theupdateframework.readthedocs.io/en/latest/INSTALLATION.html#install-for-development) the dev requirements.

That said, if you feel you are done with this PR, let me know: I can continue this as well later this week since the latest issue is really mine.

jku committed:
RequestsFetcher: satisfy mypy with small fix

The typeshed annotations for requests say that the hostname could be None:
I think this is untrue but let's keep mypy happy.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
There's a fix in https://github.com/jku/python-tuf/commits/netloc -- I think I'm not currently allowed to push the fix to this branch but options are:
* I can make a new PR with your commit and mine
* You can include that fix in this PR
* you can check _Allow edits and access to secrets by maintainers_ in this PR so I can push the fix here

any way works for me

jku reviewed (approved):
LGTM: the empty hostname leads to a session per scheme, so e.g. all "file://" urls will use the same sessions
